plugin/fractional_rewrite/setup.go

```diff
@@ -0,0 +1,70 @@
+package fractional_rewrite
+
+import (
+	"context"
+	"math/rand"
+	"strconv"
+	"strings"
+
+	"github.com/coredns/caddy"
+	"github.com/coredns/coredns/core/dnsserver"
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/request"
+)
+
+func init() { plugin.Register("fractional_rewrite", setup) }
+
+func setup(c *caddy.Controller) error {
+	rewrite, err := fractionalRewriteParse(c)
+	if err != nil {
+		return plugin.Error("fractional_rewrite", err)
+	}
+
+	dnsserver.GetConfig(c).AddPlugin(func(next plugin.Handler) plugin.Handler {
+		return fractionalRewrite{Next: next, Rule: rewrite}
+	})
+
+	return nil
+}
+
+func fractionalRewriteParse(c *caddy.Controller) (Rule, error) {
+	c.Next()
+	args := c.RemainingArgs()
+	if len(args) != 4 {
+		return nil, plugin.Error("fractional_rewrite", c.ArgErr())
+	}
+	ruleName := args[0]
+	fraction, err := strconv.ParseFloat(args[1], 64)
+	if err != nil {
+		return nil, plugin.Error("fractional_rewrite", c.ArgErr())
+	}
+	switch ruleName {
+	case "suffix":
+		return &suffixRule{
+			args[2],
+			args[3],
+			fraction,
+		}, nil
+	default:
+		return nil, plugin.Error("fractional_rewrite", c.ArgErr())
+	}
+}
+
+type Rule interface {
+	// Rewrite rewrites the current request.
+	Rewrite(ctx context.Context, state request.Request)
+}
+
+type suffixRule struct {
+	Suffix      string
+	Replacement string
+	Fraction    float64
+}
+
+// Rewrite rewrites the current request when the name ends with the matching string.
+func (rule *suffixRule) Rewrite(ctx context.Context, state request.Request) {
+	r := rand.Float64()
+	if strings.HasSuffix(state.Name(), rule.Suffix) && r < rule.Fraction {
```

Comment:
Have you considered splitting coredns instances between those using canary and those using prod, rather than splitting in the request path? This could be done on plugin initialisation rather than in the request path.

I see two main advantages to doing so:

1. In normal operations, if canary and prod return different endpoint sets (for example because we just implemented a new feature in tina that selects a different set of endpoints), clients may be switching between those results on each cache expiry, which may perturb client behaviours (connection churn, etc.).
2. It improves failure isolation. If there is a problem with canaries (e.g. responses crashing coreDNS), then all Tina clients are will be affected at a fraction of their DNS traffic. The result would be much more problematic than if only a fix subset of clients are affected.

Reply:
Thanks! I think these are valid concerns.
1 Bind clients with either canary/prod - I think this can be done by adopting a better hashing algorithm. Instead of doing `rand.Float64()`, we map/hash `client_id`(or something similar) to a deterministic int.
2 response-related failure isolation - it's hard to achieve at plugin level. Adding coreDNS installations can address this(I am assuming the strategy is to have coreDNS canary that only routes to tina-C, ratio of coreDNS-C and coreDNS-P controls how much traffic goes into canary), but it doesn't really solve the client-to-canary binding problem, correct?

Reply:
Agreed on #1, although I thought you could also do it at client initialization by coreDNS instances to either canary or not canary. Basing it on client id (if we have such a thing) would work, too.

I didn't fully grasp your point #2. My point was that by just mapping client id to be deterministically going to canary or not, 2 become less of an issue: only a subset of clients may be affected by an issue in canary. Does that make sense?

Reply:
I don't think that splitting coredns instances will help with either #1 or #2 unless we add a second coredns service and configure a subset of node-local DNS to use the second coredns service. Otherwise, the requests will be randomly distributed between 2 types of coredns instances and we are going to have exactly the same problems.

I also see problems with consistent hashing approach. If we use client IP as a hash key (which I believe is the only client-specific identifier that we have at our disposal) If we have a particular client IP, how can we deterministically say which option (canary or non-canary) we are going to use for this specific client? We could hash the IP to a smaller range, for example [1, 100] and say that if `hash <= 10` -> use canary. the problem with this is that IP distribution isn't uniform, so no matter which hashing algorithm we choose we may end up in a situation when the canary is used for way more than 10% of requests or not used at all.

Another option to implement consistent hashing that I can think of is to maintain the list of clients and mark some clients as `canary` This affects performance because now we are going to maintain the list of connections, we must protect this list by a lock and do a search in this list for every request. Given the number of requests CoreDNS is processing this could be a problem. It is also not clear how to evict items from this list.

Reply:
hmm.. agree on using `client_ip`(I wasn't able to find a better key than this, it doesn't seem we can distinguish client/pod from a dns.msg + response writer). By IP distribution not being uniform, do you mean the number of pods running under a service varies or k8s assigns IP to services using some non-uniform algorithm? If it's the 1st case, I think if we look at the release at service level, it's still ~10% services going to canary. The real distribution will depend on service-to-pod distribution, I don't expect it to be deviate from 10% a lot.

Reply:
>  k8s assigns IP to services using some non-uniform algorithm

This is definitely the case. The full theoretical range of IP v4 is [0, 2^32] and k8s assignes Ips sequentially per node from a given CIDRs. If you use some simple hash function, like `ip mod 10` you will likely end up with way more items with hash 0 or 1 than 9.

Reply:
OK, I think I am missing some pieces of information - for some reason I assumed a CoreDNS instance was running per pod or per node, but reading more about it I guess it's not the case. Would someone be able to explain me how CoreDNS is set up? Now just poking around some docs (e.g. https://datadoghq.atlassian.net/wiki/spaces/FABRIC/pages/1787593762/Consul+CoreDNS) I think it's a few instances per cluster?

> I also see problems with consistent hashing approach. If we use client IP as a hash key (which I believe is the only client-specific identifier that we have at our disposal) If we have a particular client IP, how can we deterministically say which option (canary or non-canary) we are going to use for this specific client?

Do we need to say in advance whether it's going to use canary or not? I think hashing the full IP address with something that gives a random distribution would be good enough. The apparent downside is that clients do not know in advance whether a given pod is going to go 100% canary or 100% prod. I still think it's better than potentially flapping between canary and prod on each request.

Reply:
I don't think the client/pod needs to know in advance if it's going to canary or not(I cannot think of obvious benefits for them to have this knowledge), plus IP changes (e.g. after redeploy for headless services I think), this approach doesn't enforce a particular set of services to always go to canary while others don't.

Reply:
I wonder whether consistent hashing for clients is really the right approach. I think there are pros/cons, but without any control over _which_ clients get 100% canary or 100% prod, we might end up fully breaking critical services if their pods are 100% set to a broken canary, whereas with fully random routing we would only partially degrade 10% of overall requests for these services. 

If the metrics are tagged properly we should still be able to determine, for example, 80% of Canary requests are failing but 100% of Prod requests are succeeding.

Reply:
Ah, I see. The tradeoff is that a broken canary potentially taking down critical services v.s. having global (light?) effect on all internal services. I cannot justify which scenario is worse, would it be Okay to implement both approaches and test them in a Gameday maybe?

Reply:
I think that's a great idea. 👍